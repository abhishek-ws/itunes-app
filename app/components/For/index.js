/**
 *
 * For
 *
 */

import React from 'react';
import Proptypes from 'prop-types';
import styled from 'styled-components';

const FlexContainer = styled.div`
  display: flex;
  flex-direction: ${(props) => (props.isrow ? `row` : `column`)};
`;

export function For({
  of,
  ParentComponent = (props) => <FlexContainer isrow {...props} />,
  renderItem,
  noParent,
  ...props
}) {
  const list = () => of.map((item, index) => ({ ...renderItem(item, index), key: index }));
  const children = () => (
    <ParentComponent {...props} data-testid="for">
      {list()}
    </ParentComponent>
  );
  if (noParent) {
    return (of || []).length ? list() : null;
  }
  return (of || []).length ? children() : null;
}

For.propTypes = {
  of: Proptypes.array,
  type: Proptypes.node,
  parent: Proptypes.object,
  iterate: Proptypes.string,
  renderItem: Proptypes.func.isRequired,
  noParent: Proptypes.bool,
  isrow: Proptypes.bool
};

export default For;
