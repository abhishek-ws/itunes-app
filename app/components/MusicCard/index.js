/**
 *
 * MusicCard
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card, Image, Typography, Button } from 'antd';
import { PauseCircleTwoTone, PlayCircleTwoTone, StopTwoTone } from '@ant-design/icons';

const { Paragraph, Title } = Typography;

const Container = styled(Card)`
  && {
    max-width: 300px;
    height: 500px;
    background-color: darkgray;
    color: white;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: flex-start;
  }
`;

const IconsContainer = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  justify-content: space-between;
`;

const CommonIconStyle = {
  fontSize: 18
};

const CommonButtonStyle = {
  display: 'flex',
  margin: 'auto 3px',
  justifyContent: 'space-between',
  alignItems: 'center',
  backgroundColor: 'transparent'
};
// const CommonIconSize = 10;

export function MusicCard({ song }) {
  const { shortDescription, artworkUrl100, artistName } = song;
  return (
    // <div data-testid="music-card">
    <Container data-testid="music-card">
      <Image src={artworkUrl100} width="80%" preview="false" height="80%" />
      <Title style={{ fontSize: 18 }} italic={true}>
        {artistName}
      </Title>
      <Paragraph style={{ fontSize: 12, height: '50px' }}>
        {shortDescription ? shortDescription : 'No Description available'}
      </Paragraph>
      <IconsContainer>
        <Button type="ghost" icon={<PlayCircleTwoTone style={CommonIconStyle} />} style={CommonButtonStyle}>
          Play
        </Button>
        <Button
          type="ghost"
          disabled={true}
          icon={<PauseCircleTwoTone style={CommonIconStyle} />}
          size="large"
          style={CommonButtonStyle}
        >
          Pause
        </Button>
        <Button type="ghost" icon={<StopTwoTone style={CommonIconStyle} />} style={CommonButtonStyle} size="large">
          Stop
        </Button>
      </IconsContainer>
    </Container>
    // </div>
  );
}

MusicCard.propTypes = {
  song: PropTypes.object
};
export default MusicCard;
