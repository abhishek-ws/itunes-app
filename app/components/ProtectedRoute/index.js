/**
 *
 * ProtectedRoute
 *
 */

import React from 'react';
import { Redirect, Route } from 'react-router-dom';
import PropTypes from 'prop-types';
import routeConstants from '@utils/routeConstants';

const ProtectedRoute = ({ isUserVerified, render: C, isLoggedIn, handleLogout, ...rest }) => {
  const isUnprotectedRoute =
    Object.keys(routeConstants)
      .filter((key) => !routeConstants[key].isProtected)
      .map((key) => routeConstants[key].route)
      .includes(rest.path) && rest.exact;

  const guardedRedirection = (to, renderProps) => {
    if (rest.path === to) {
      return <C {...renderProps} />;
    }
    return <Redirect to={to} />;
  };

  const handleRedirection = (renderProps) => {
    let to;
    if (!isLoggedIn) {
      // user is not logged in
      if (!isUnprotectedRoute) {
        to = routeConstants.login.route;
        handleLogout();
      } else {
        // not logged in and trying to access an unprotected route so don't redirect
        return <C {...renderProps} />;
      }
    } else if (isLoggedIn) {
      // user is logged in
      if (isUnprotectedRoute) {
        to = routeConstants.dashboard.route;
      } else {
        // logged in and accessing a protected route
        return <C {...renderProps} />;
      }
    }
    return guardedRedirection(to, renderProps);
  };
  return <Route {...rest} render={handleRedirection} />;
};

ProtectedRoute.propTypes = {
  render: PropTypes.any,
  isLoggedIn: PropTypes.bool,
  isUserVerified: PropTypes.bool,
  handleLogout: PropTypes.func
};

export default ProtectedRoute;
