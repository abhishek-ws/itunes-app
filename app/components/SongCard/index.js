/**
 *
 * MusicCard
 *
 */

import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card, Typography, Button } from 'antd';
import { PlayCircleTwoTone, StopTwoTone } from '@ant-design/icons';
import { colors } from '@app/themes';
import { translate } from '../IntlGlobalProvider/index';
import { Link } from 'react-router-dom';
import If from '@components/If';
import { T } from '@components/T';

const { Paragraph } = Typography;

const Container = styled(Card)`
  && {
    width: ${(props) => (props.width ? props.width : 25)}em;
    height: ${(props) => (props.height ? props.height : 32)}em;
    background-color: ${colors.songCardBg};
    display: block;
    border-radius: 2em;
    box-shadow: inset -6px -6px 12px rgba(0, 0, 0, 0.8), inset 6px 6px 12px rgba(255, 255, 255, 0.4);
    overflow-x: scroll;
    &::-webkit-scrollbar {
      display: none;
    }
  }
`;

const IconsContainer = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  justify-content: space-between;
`;

const commonIconStyle = {
  fontSize: 18
};

const ControlButton = styled(Button)`
  && {
    color: white;
    display: flex;
    margin: auto 3px;
    justify-content: space-between;
    align-items: center;
    background-color: transparent;
  }
`;

const StyledT = styled(T)`
  && {
    font-size: ${(props) => (props.trackDetails ? 2 : 1.1)}em;
    margin: ${(props) => (props.trackDetails ? '20 0' : 0)};
  }
`;
const StyledPrice = styled(T)`
  && {
    color: ${colors.styledPriceColor};
    font-size: ${(props) => (props.trackDetails ? 40 : 16)};
    margin: 20 10;
  }
`;

const StyledParagraph = styled(Paragraph)`
  && {
    margin-top: 10;
    color: ${colors.styledParaColor};
    font-size: ${(props) => (props.trackDetails ? 1.5 : 1)}em;
    height: ${(props) => (props.shortDescription ? 7 : 4.5)}em;
    overflow: hidden;
  }
`;

const HeaderFooter = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const StyledImage = styled.img`
  width: 100px;
  height: ${(props) => (props.trackDetails ? 18 : 14)}em;
  margin-bottom: 2em;
`;

export function SongCard({ song, trackDetails, width, height, onActionClick }) {
  const { trackName, trackPrice, artworkUrl100, previewUrl, trackId } = song;

  const [play, setPlay] = useState(false);
  const songElement = useRef(null);

  const actions = {
    PLAY: 'playMusic',
    STOP: 'stopMusic'
  };

  useEffect(() => {
    if (songElement.current.paused) {
      setPlay(false);
    }
  });

  const handleMusic = async (action, songUrl) => {
    switch (action) {
      case actions.PLAY:
        songElement.current.src = songUrl;
        setPlay(!play);
        onActionClick(songElement);
        await songElement.current.play();
        break;
      case actions.STOP:
        songElement.current.src = '';
        setPlay(!play);
    }
  };

  return (
    <Container width={width} height={height} data-testid="song-card">
      <HeaderFooter>
        <StyledImage src={artworkUrl100} />
        <Link to={`/details/${trackId}`}>
          <StyledT trackDetails={trackDetails} text={trackName} />
        </Link>
      </HeaderFooter>

      <StyledParagraph data-testid="para-test" trackDetails={trackDetails} shortDescription={song.shortDescription}>
        <If
          condition={song.shortDescription}
          otherwise={song.longDescription ? song.longDescription : translate('no_description')}
        >
          {song.shortDescription}
        </If>
      </StyledParagraph>
      <IconsContainer>
        <ControlButton
          data-testid="play-btn"
          onClick={() => handleMusic(actions.PLAY, previewUrl)}
          disabled={play}
          type={play ? 'text' : 'ghost'}
          icon={<PlayCircleTwoTone style={commonIconStyle} />}
        >
          {translate('play-btn')}
        </ControlButton>
        <ControlButton
          data-testid="stop-btn"
          disabled={!play}
          onClick={() => handleMusic(actions.STOP, previewUrl)}
          type={play ? 'ghost' : 'text'}
          icon={<StopTwoTone style={commonIconStyle} />}
          size="large"
        >
          {translate('stop-btn')}
        </ControlButton>
      </IconsContainer>
      <HeaderFooter>
        <If condition={trackPrice} otherwise={<StyledPrice data-testid="no-price-tag" id="no-price" />}>
          <StyledPrice id="track-price" values={{ trackPrice }} />
        </If>
      </HeaderFooter>
      <audio data-testid="audio-element" ref={songElement}></audio>
    </Container>
  );
}

SongCard.propTypes = {
  song: PropTypes.object,
  width: PropTypes.number,
  height: PropTypes.number,
  trackDetails: PropTypes.bool,
  onActionClick: PropTypes.func
};
export default SongCard;
