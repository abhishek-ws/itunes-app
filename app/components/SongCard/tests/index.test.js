/**
 *
 * Tests for MusicCard
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { renderWithIntl, timeout } from '@utils/testUtils';
import SongCard from '../index';
import { setIntl, translate } from '@app/components/IntlGlobalProvider/';
import getIntl from '@utils/createIntl';

describe('<SongCard/> tests', () => {
  beforeAll(() => {
    setIntl(getIntl());
  });
  const song = {
    trackName: 'Track',
    trackPrice: 200,
    trackId: 1234,
    shortDescription: 'Song Description',
    artworkUrl100:
      'https://www.google.com/url?sa=i&url=https%3A%2F%2Fdribbble.com%2Fshots%2F14395014-Music-Logo&psig=AOvVaw3I_T0J8_ZRKW_g4VF_KmKz&ust=1630933127025000&source=images&cd=vfe&ved=2ahUKEwiT5uO-8efyAhWVHLcAHVSbCvQQjRx6BAgAEAk',
    artistName: 'Artist Name',
    previewUrl: 'https://mockurl.com/'
  };

  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<SongCard song={song} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 SongCard component', () => {
    const { getAllByTestId } = renderWithIntl(<SongCard song={song} />);
    expect(getAllByTestId('song-card').length).toBe(1);
  });

  it('should render No-price translation when trackPrice is not available', () => {
    const song = {
      trackName: 'Track',
      trackPrice: null,
      trackId: 1234,
      shortDescription: 'Song Description',
      artworkUrl100:
        'https://www.google.com/url?sa=i&url=https%3A%2F%2Fdribbble.com%2Fshots%2F14395014-Music-Logo&psig=AOvVaw3I_T0J8_ZRKW_g4VF_KmKz&ust=1630933127025000&source=images&cd=vfe&ved=2ahUKEwiT5uO-8efyAhWVHLcAHVSbCvQQjRx6BAgAEAk',
      artistName: 'Artist Name',
      previewUrl: 'https://mockurl.com/'
    };

    const { getByTestId } = renderWithIntl(<SongCard song={song} />);
    expect(getByTestId('no-price-tag')).toHaveTextContent(translate('no-price'));
  });

  it('should render the song`s short description accordingly when rendering the card on Home Page / Grid from given from song prop', () => {
    const song = {
      trackName: 'Track',
      trackPrice: 200,
      trackId: 1234,
      shortDescription: 'SongShortDescription',
      artworkUrl100:
        'https://www.google.com/url?sa=i&url=https%3A%2F%2Fdribbble.com%2Fshots%2F14395014-Music-Logo&psig=AOvVaw3I_T0J8_ZRKW_g4VF_KmKz&ust=1630933127025000&source=images&cd=vfe&ved=2ahUKEwiT5uO-8efyAhWVHLcAHVSbCvQQjRx6BAgAEAk',
      artistName: 'Artist Name',
      previewUrl: 'https://mockurl.com/'
    };

    const { getByTestId } = renderWithIntl(<SongCard song={song} />);
    expect(getByTestId('para-test')).toHaveTextContent(song.shortDescription);
  });

  it('should render the song`s long description accordingly when rendering the card on Track Details Page from given from song prop', () => {
    const song = {
      trackName: 'Track',
      trackPrice: 200,
      trackId: 1234,
      longDescription: 'LongShortDescription',
      artworkUrl100:
        'https://www.google.com/url?sa=i&url=https%3A%2F%2Fdribbble.com%2Fshots%2F14395014-Music-Logo&psig=AOvVaw3I_T0J8_ZRKW_g4VF_KmKz&ust=1630933127025000&source=images&cd=vfe&ved=2ahUKEwiT5uO-8efyAhWVHLcAHVSbCvQQjRx6BAgAEAk',
      artistName: 'Artist Name',
      previewUrl: 'https://mockurl.com/'
    };

    const { getByTestId } = renderWithIntl(<SongCard song={song} />);
    expect(getByTestId('para-test')).toHaveTextContent(song.longDescription);
  });

  it('should disable-Enable the Play-Stop buttons once, song is played or stopped accordingly', async () => {
    let audio;

    const playFn = jest.spyOn(window.HTMLMediaElement.prototype, 'play').mockImplementation(() => {
      audio.paused = !audio.paused;
    });
    const pauseFn = jest.spyOn(window.HTMLMediaElement.prototype, 'pause').mockImplementation(() => {
      audio.paused = !audio.paused;
    });

    const { getByTestId } = renderWithIntl(<SongCard song={song} onActionClick={jest.fn()} />);
    audio = getByTestId('audio-element');
    fireEvent.click(getByTestId('play-btn'));

    await timeout(500);
    expect(getByTestId('play-btn')).toBeDisabled();
    expect(getByTestId('stop-btn')).toBeEnabled();

    expect(playFn).toBeCalled();
    expect(pauseFn).not.toBeCalled();

    fireEvent.click(getByTestId('stop-btn'));

    await timeout(500);

    expect(pauseFn).toBeCalled();
    expect(getByTestId('play-btn')).toBeEnabled();
    expect(getByTestId('stop-btn')).toBeDisabled();
  });
});
