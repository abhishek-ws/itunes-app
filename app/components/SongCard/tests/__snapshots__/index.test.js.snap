// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SongCard/> tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered SongCard__Container-sc-163fh4j-0 rEpnk"
      data-testid="song-card"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="SongCard__HeaderFooter-sc-163fh4j-6 bBcspZ"
        >
          <img
            class="SongCard__StyledImage-sc-163fh4j-7 egedQz"
            src="https://www.google.com/url?sa=i&url=https%3A%2F%2Fdribbble.com%2Fshots%2F14395014-Music-Logo&psig=AOvVaw3I_T0J8_ZRKW_g4VF_KmKz&ust=1630933127025000&source=images&cd=vfe&ved=2ahUKEwiT5uO-8efyAhWVHLcAHVSbCvQQjRx6BAgAEAk"
          />
          <a
            href="/details/1234"
          >
            <p
              class="T__StyledText-gjlic1-0 kYcyRq SongCard__StyledT-sc-163fh4j-3 gjIxDk"
              data-testid="t"
            >
              Track
            </p>
          </a>

        </div>
        <div
          class="ant-typography SongCard__StyledParagraph-sc-163fh4j-5 ctnAOd"
          data-testid="para-test"
          shortdescription="Song Description"
        >
          Song Description
        </div>
        <div
          class="SongCard__IconsContainer-sc-163fh4j-1 fJrksM"
        >
          <button
            class="ant-btn ant-btn-ghost SongCard__ControlButton-sc-163fh4j-2 dcxDWV"
            data-testid="play-btn"
            type="button"
          >
            <span
              aria-label="play-circle"
              class="anticon anticon-play-circle"
              role="img"
              style="font-size: 18px;"
            >
              <svg
                aria-hidden="true"
                data-icon="play-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  fill="#1890ff"
                />
                <path
                  d="M512 140c-205.4 0-372 166.6-372 372s166.6 372 372 372 372-166.6 372-372-166.6-372-372-372zm164.1 378.2L457.7 677.1a8.02 8.02 0 01-12.7-6.5V353a8 8 0 0112.7-6.5l218.4 158.8a7.9 7.9 0 010 12.9z"
                  fill="#e6f7ff"
                />
                <path
                  d="M676.1 505.3L457.7 346.5A8 8 0 00445 353v317.6a8.02 8.02 0 0012.7 6.5l218.4-158.9a7.9 7.9 0 000-12.9z"
                  fill="#1890ff"
                />
              </svg>
            </span>
            <span>
              Play
            </span>
          </button>
          <button
            class="ant-btn ant-btn-text ant-btn-lg SongCard__ControlButton-sc-163fh4j-2 dcxDWV"
            data-testid="stop-btn"
            disabled=""
            type="button"
          >
            <span
              aria-label="stop"
              class="anticon anticon-stop"
              role="img"
              style="font-size: 18px;"
            >
              <svg
                aria-hidden="true"
                data-icon="stop"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm288.5 682.8L277.7 224C258 240 240 258 224 277.7l522.8 522.8C682.8 852.7 601 884 512 884c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372c0 89-31.3 170.8-83.5 234.8z"
                  fill="#1890ff"
                />
                <path
                  d="M512 140c-205.4 0-372 166.6-372 372s166.6 372 372 372c89 0 170.8-31.3 234.8-83.5L224 277.7c16-19.7 34-37.7 53.7-53.7l522.8 522.8C852.7 682.8 884 601 884 512c0-205.4-166.6-372-372-372z"
                  fill="#e6f7ff"
                />
              </svg>
            </span>
            <span>
              Stop
            </span>
          </button>
        </div>
        <div
          class="SongCard__HeaderFooter-sc-163fh4j-6 bBcspZ"
        >
          <p
            class="T__StyledText-gjlic1-0 kYcyRq SongCard__StyledPrice-sc-163fh4j-4 zUpYc"
            data-testid="no-price-tag"
          >
            no-price
          </p>
        </div>
        <audio
          data-testid="audio-element"
        />
      </div>
    </div>
  </div>
</body>
`;
